### PR TITLE
feat: generics for useMapState

### LINKED ISSUE #585: Use generics in hooks like useMapState
Hooks that deal with and return data should be using generics to give users the benefit of type analysis on their data.

### PR DESCRIPTION
**Pull request by qqpann**
Added generics.

---
close #585

### CODE REVIEW

File: src/hooks/useMapState.ts [hunk 1]
```diff
@@ -6,82 +6,81 @@ import { useCallback, useState } from "react";
...
-  any,
+function useMapState<T>(initialValue: T): [
+  T,
   {
-    set: (key: any, value: any) => void;
-    has: (key: any) => boolean;
-    setMultiple: (...keys: any[]) => void;
-    remove: (key: any) => void;
-    removeMultiple: (...keys: any[]) => void;
+    has: (key: keyof T) => boolean;
```
Comment (imbhargav5) [inline src/hooks/useMapState.ts:12]: @qqpann Nice. Let's alias this into a type at the top of this file and use it here for good readibility.

File: src/hooks/useMapState.ts [hunk 2]
```diff
@@ -6,82 +6,81 @@ import { useCallback, useState } from "react";
...
-      });
+      }
 
       return newMap;
     });
   }, []);
 
   const removeAll = useCallback(() => {
-    setMap({});
+    setMap({} as T);
```
Comment (imbhargav5) [inline src/hooks/useMapState.ts:70]: @qqpann Same thing about type assertions.

File: src/hooks/useMapState.ts [hunk 3]
```diff
@@ -6,82 +6,81 @@ import { useCallback, useState } from "react";
...
       return newMap;
     });
   }, []);
 
-  const remove = useCallback((key: any) => {
+  const remove = useCallback((key: keyof T) => {
     setMap((currentMap) => {
-      const newMap = {};
-      Object.keys(currentMap).forEach((mapKey) => {
+      const newMap = {} as T;
```
Comment (imbhargav5) [inline src/hooks/useMapState.ts:58]: @qqpann Same thing about type assertions.

**Comment by imbhargav5** [comment]
@qqpann So the breaking changes are typescript related and not JS related right?

**Comment by qqpann** [comment]
> So the breaking changes are typescript related and not JS related right?

Yes. I didn't confirm but I believe it will work as it used to in JS

File: src/hooks/useMapState.ts (hunk 1 above)
Comment (qqpann) [reply src/hooks/useMapState.ts]: I changed it. With this change, you have to specify the second type when explicitly specifying genetics

File: src/hooks/useMapState.ts [hunk 4]
```diff
@@ -6,82 +6,81 @@ import { useCallback, useState } from "react";
...
     }));
   }, []);
 
-  const removeMultiple = useCallback((...keys) => {
+  const removeMultiple = useCallback((...keys: Array<keyof T>) => {
     setMap((currentMap) => {
-      const newMap = {};
-      Object.keys(currentMap).forEach((key) => {
-        if (!keys.includes(key)) {
+      const newMap = {} as T;
```
[the remark this answers is not in the document]
Comment (qqpann) [reply src/hooks/useMapState.ts]: Fixed it by making more changes. Although there came out more eslint errors, it passed the test and build.

**Comment by imbhargav5** [comment]
@qqpann I might take a week to merge this just to be sure. Giving you a headsup.

**Pull request merged**